database/src/main/java/com/firebase/ui/database/FirebaseArray.java

```diff
@@ -26,69 +28,72 @@
  * This class implements an array-like collection on top of a Firebase location.
  */
 class FirebaseArray implements ChildEventListener {
-    public interface OnChangedListener {
+    public interface ChangeEventListener {
         enum EventType {ADDED, CHANGED, REMOVED, MOVED}
 
         void onChanged(EventType type, int index, int oldIndex);
 
-        void onCancelled(DatabaseError databaseError);
+        void onCancelled(DatabaseError error);
     }
 
+    protected ChangeEventListener mListener;
+    protected boolean mIsListening;
     private Query mQuery;
-    private OnChangedListener mListener;
     private List<DataSnapshot> mSnapshots = new ArrayList<>();
 
     public FirebaseArray(Query ref) {
         mQuery = ref;
+    }
+
+    public void setChangeEventListener(@NonNull ChangeEventListener listener) {
+        if (mIsListening && listener == null) {
+            throw new IllegalStateException("Listener cannot be null.");
+        }
+        mListener = listener;
+    }
+
+    public void startListening() {
+        if (mListener == null) throw new IllegalStateException("Listener cannot be null.");
         mQuery.addChildEventListener(this);
+        mIsListening = true;
     }
 
-    public void cleanup() {
+    public void stopListening() {
         mQuery.removeEventListener(this);
+        mSnapshots.clear();
+        mIsListening = false;
     }
 
-    public int getCount() {
+    public int size() {
```

Comment:
@TheCraftKid Nope, these aren't breaking api changes. 🎉 We're really lucky @puf wanted to hold out on open sourcing `FirebaseArray` and `FirebaseIndexArray` or these would have been (You're thinking of `FirebaseRecyclerViewAdapter#getItemCount()`, but I didn't change that.)

As for enums, @puf decided against it and I'm fine with that because I found out proguard usually converts enums to `int`s anyway: http://stackoverflow.com/a/32236950/4548500. Cheers!